desc/perturbations.py

```diff
@@ -205,81 +202,75 @@ def perturb(  # noqa: C901 - FIXME: break this up into simpler pieces
     dx2_reduced = jnp.zeros_like(x_reduced)
     dx3_reduced = jnp.zeros_like(x_reduced)
 
+    xz = objective.unpack_state(jnp.zeros_like(x), False)[0]
     # tangent vectors
-    tangents = jnp.zeros((objective.dim_x,))
+    tangents = jnp.zeros((eq.dim_x,))
     if "Rb_lmn" in deltas.keys():
         con = get_instance(constraints, BoundaryRSelfConsistency)
-        A = con.derivatives["jac_unscaled"]["R_lmn"](
-            *[jnp.zeros(con.dimensions[arg]) for arg in con.args]
-        )
+        A = con.jac_unscaled(xz)["R_lmn"]
```

Comment:
In terms of your perturbation-continuation paper, is A = df/dR_lmn? Similarly, what is "tangents"?

Reply:
Not quite. This `A` is the jacobian of the constraint that enforces consistency between the Fourier-Zernike basis in the volume and the double Fourier basis on the boundary. That constraint can be written as `A*R_lmn = Rb_mn`. When we perturb the boundary, we need to translate that into a perturbation of the volume. In the paper notation we have $\partial f/\partial c \Delta c$ where c are the things we're perturbing, and that's basically what we're doing here.

$$\frac{\partial f}{\partial c}\Delta c = \frac{\partial f}{\partial R_{lmn}}\frac{\partial R_{lmn}}{\partial Rb_{mn}}\Delta c = \frac{\partial f}{\partial R_{lmn}}A \Delta c $$

`tangents` is basically just $A \Delta c$, which is a tangent vector in the parlance of differential geometry and JAX